design-proposals/volume-claim-templates.md

```diff
@@ -0,0 +1,117 @@
+
+# Design
+
+A new field will be added to the VM object called **VolumeClaimTemplates**. This field will provide the user a way to define and create a PVC owned by the VM while providing partial/full PVC spec embedded in the VM spec. In case of partial spec we will make sure the preferred VolumeMode and AccessModes combination according to the StorageProfile associated with provided/default StorageClass are used.  
+This field should replace the DataVolumeTemplates field. Instead of providing a template for a DataVolume which eventually creates one, we will provide a PVC template and create it. DataVolumes and DataVolumeTemplates should be deprecated eventually.  
+A new API will be created in the CDI API library which will receive a partial PVC template and using existing CDI logic will return a complete PVC spec.  
+The provided PVC template in the VolumeClaimTemplate should have a source that can be handled by the new existing CDI volume populators API (DataSourceRef with CDI CR source) or a standalone k8s way (user costume populator, empty disk etc..)
+
+A virtctl command called `create-pvc` can be added. Given minimal flags such as name, size and source this command can generate and create a PVC. The command will use the CDI API similarly as will be done in the VM controller.
+
+## API Examples
+### VM with VolumeClaimTemplates, VolumeImportSource as volume source
+
+This example will create a VolumeImportSource along side with a VM. Once the VM is created we will generate a PVC spec with the optimal VolumeMode and AccessModes. This PVC will be created and populated with the provided image in the VolumeImportSource url.
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportSource
+metadata:
+  name: my-import-source
+spec:
+  source:
+      http:
+         url: "https://download.cirros-cloud.net/0.4.0/cirros-0.4.0-x86_64-disk.img"
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  running: true
+  volumeClaimTemplates:
+  - metadata:
+      name: cirros-pvc
+    spec:
+    dataSourceRef:
+      apiGroup: cdi.kubevirt.io
+      kind: VolumeImportSource
+      name: my-import-source
+    resources:
+      requests:
+        storage: 1Gi
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+      - persistentVolumeClaim:
+          name: cirros-pvc
+        name: cirros-disk
+```
+
+#### Open question:
+With datavolumestemplate one of the downsides was that we couldn't use generateName of the datavolume name.
+It will be cool if we could use it for the PVC name. Issue: we need to mention the PVC name in the vm.spec.template.volumes part when writing the VM spec. Any idea how to go past that? update the PVC name there after the PVC is created sounds like not the best solution. WDYT?
```

Comment:
Since Disks are already automatically generated for Volumes, we could automatically create Volumes for each VolumeClaimTemplate and fill in the name? Maybe only in the resulting VMI? This would help keeping the VM's manifest smaller too. Right now it looks like duplicated information to me.

Reply:
sounds pretty reasonable to me, after creating the PVC and before creating the VMI to add a volume to the volumes list with the generated PVC name, then it will also exist in the resulting VMI.

Reply:
That isn't going to work, we need the list of volumes to be present during VMI submission for the mutation webhook to populate any missing disks:

https://github.com/kubevirt/kubevirt/blob/f587faf23ba8462c8d61274b81aebdeedfbe8058/pkg/util/util.go#L246-L267

AFAICT we are talking about only creating these resources *after* submission right?

Reply:
we are talking on a time after creating the VM, going over the volumeClaimTemplate, creating the PVC. Then adding to the volumes list in the vm.spec.template.spec.volumes the volume with the generated name. Only then will be proceed to creating the VMI with the already updated template, so it should exist in the VMI submission.

Reply:
Apologies I for some reason thought `volumeClaimTemplate` was under the VMI spec in this design.

I would however note that @rmohr recently highlighted to me that mutating the `VirtualMachine` post submission like this is not great for declarative management systems. We currently do this for instance types when capturing a `ControllerRevision` and are tracking the potential removal of this patching and move to storing it somewhere under `.Status` in the following bug. https://github.com/kubevirt/kubevirt/issues/10145

Reply:
I would prefer that VM volumes remain explicit

Reply:
DataVolumes cannot support `generateName` reliably.  There are "reasons" but they are not applicable to this design.

Reply:
@mhenriks When having a volumeClaimTemplate, what's the benefit of specifying it again as volume?

Reply:
This would result in a PVC being owned by the VM but left unused? Sounds strange to me, is it really that common?

Reply:
I doubt it's common but it's not unreasonable.  Main motivation is that I feel that we should closely follow the `StatefulSet` model

Reply:
One more comment: `StatefulSet` allows to specify multiple containers, while `VirtualMachine` creates just one possible consumer. So the `StatefulSet` API needs to have some freedom in order to allow assigning the volume to container(s). On 'VirtualMachine` there is no discrepancy on which consumer the volume should be assigned to. WDYT?

Reply:
I am still still strongly in favor of user supplying the volumes explicitly.  Also, don't forget that volumes are referenced in the domain section and ordering there is important.  We certainly cannot fill that in automatically

Reply:
AFAIK Volumes are used in the Disks section of the domain spec. These are already filled in automatically if not present. Are they used somewhere else? Also I would not drop the volumes section entirely, just fill in Voulems automatically if possible but still allow users to specify Volumes themselves.

Reply:
After talking with @mhenriks there is a good solutions of having PVC name unique and random by making it dependent on the VM name. If the VM name is generated then the PVC name when the VMI is created will appear in the VMI as <VM-(generated)-name>-<pvc-name> similar to how VMI name gets the generated name after the VM is created. I will add this to the design doc with example to make it clearer, currently this sounds like the best solution.

Reply:
I agree with @0xFelix that generating `Volumes` and later `Disks` from a list of `VolumeClaimTemplates` during VMI mutation would be nice and still possible with the use of `generateName` within these templates. Ordering would just be inherited from the `VolumeClaimTemplates` list as it is currently when populating missing `Disks` from `Volumes`. All we would need is the VM controller to stash the generated names in the status of the VM for use later by the VMI mutation webhook I guess?

Reply:
How about when volumeclaimTemplates are used in combination with other volumes like containerdisks?  How will ordering be resolved in these cases?

Reply:
I see this as an optimization we can discuss in a separate design proposal/pr.